pkg/cmd/search/shared/qualifier.go

```diff
@@ -0,0 +1,140 @@
+package shared
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// These regexs are not perfect and mostly used to give some input validation.
+// They are more permissive than the server so we can provide early feedback
+// to the user in most cases but there are some inputs that will pass the
+// regexs and then be rejected by the server.
+var rangeRE = regexp.MustCompile(`^(>|>=|<|<=|\*\.\.)?\d+(\.\.(\*|\d+))?$`)
+var dateTime = `(\d|-|\+|:|T)+`
+var dateTimeRangeRE = regexp.MustCompile(fmt.Sprintf(`^(>|>=|<|<=|\*\.\.)?%s(\.\.(\*|%s))?$`, dateTime, dateTime))
+
+type validator func(string) error
+
+type qualifier struct {
+	key       string
+	kind      string
+	set       bool
+	validator validator
+	value     string
+}
+
+type parameter = qualifier
+
+func NewQualifier(key, kind, value string, validator func(string) error) *qualifier {
+	return &qualifier{
+		key:       key,
+		kind:      kind,
+		validator: validator,
+		value:     value,
+	}
+}
+
+func NewParameter(key, kind, value string, validator func(string) error) *parameter {
+	return &parameter{
+		key:       key,
+		kind:      kind,
+		validator: validator,
+		value:     value,
+	}
+}
+
+func (q *qualifier) IsSet() bool {
+	return q.set
+}
+
+func (q *qualifier) Key() string {
+	return q.key
+}
+
+func (q *qualifier) Set(v string) error {
+	if q.validator != nil {
+		err := q.validator(v)
+		if err != nil {
+			return err
+		}
+	}
+	q.set = true
+	q.value = v
+	return nil
+}
+
+func (q *qualifier) String() string {
+	if strings.ContainsAny(q.value, " \"\t\r\n") {
+		return fmt.Sprintf("%q", q.value)
+	}
```

Comment:
I was unsure if it should be up to the qualifier to quote itself incase there is whitespace, seems like it could easily fit within the `searcher` responsibilities but this solution seemed clean as well. I could be convinced to move this functionality to `searcher` if people have strong feelings.